### ISSUE TITLE
Add option to define a max datapoints per query rule

### ISSUE THREAD

**Issue by replay** [raintank/metrictank#573]
We want to be able to define that a query should never be replied with more than a given number of data points. To achieve this we'll need to lookup the aggregation schemas of the queried targets and compare their number of data points per time with the queried time range.
If we see that the result would contain too many data points we should switch to the next less granular aggregation and apply the same test again. If there is no next less granular aggregation the query should be rejected fast, without further processing.
This is similar to `max-points-per-req` but on a per target basis.

**Comment by woodsaj** [comment]
so i think we should just remove the existing request limits and replace with a simple lowWaterMark and highWaterMark for total points per requests.

We select lower resolution rollups to try and keep the total points below the lowWaterMark.  Queries that are above the highWaterMark are rejected as being too large.

eg. assuming lowWaterMark=1million and highWaterMark=50million, and a retention policy of 1s:8d,1m:30day

requesting the last 7days for 10 series, (60*60*24*7*10=6048000 raw points) would result in the query using the 1min rollups ( 60*24*7*10=100800 points )

but requesting the last 7days for 5000 series would be rejected as even when using the lowest resolution data the number of datapoints exceeds 50million.  60*24*7*5000=50400000

**Comment by replay** [comment]
If somebody is looking a query that includes 5 targets, then he removes 4 of the 5 targets and looks at the one that's left, then it would be weird if the resolution of that target changes simply because he removed some other targets.

so i think the low / high watermark should rather be per target, but additionally we can have a limit for the entire request to limit cases where many targets are queried...

**Comment by DanCech** [comment]
good point @replay, I agree

**Comment by woodsaj** [comment]
while that is a good point, the purpose of the limits are to constrain the "cost" of the request, so it only makes sense to apply the limits based on the total request. 

allowing users to use high resolution data when a query spans a few series but force the use lower resolution data when 1000's of series are needed is the intended behavior.

If you change the the complexity of your query, then it is not unreasonable to see a change in the resolution.

**Comment by Dieterbe** [comment]
The two tunables we currently have in place (`max-points-per-req` and `max-days-per-req`) seem very ineffective at this point. They look at number of targets requested, not number of series.  I think this dates from before we supported query patterns (back when 1 target = 1 serie requested = 1 serie returned). now that a single target can result in arbitrarily many series, this doesn't make sense anymore. 
The fact that the latter is expressed in time instead of number of points has pro's and cons I think:
* con - if you want to use this limit as a datapoints limit, you have to think about your rollups and retentions and converting points to a time range
* pro - very easy to understand

I think we have to take a step back and first determine/agree which parameters control cost / resource utilisation and determine where we want to place limits. We don't have to go too extreme here and back up everything with scientific measurements though.

For example, we know that decoding chunks is pretty expensive (when runtime consolidating high resolution data, decoding actually takes quite a bit more time).  How about loading data from cassandra, we know the less reads from cassandra the better but is that because our current cassandrastore code sucks (#350), is it due to the number of seeks, or due to the size of the data being loaded? how should we really determine the impact of a given query on cassandra load? should we look at number of chunks that need to be requested instead of number of points? 

Having the chunkcache changes this dynamic but I don't think we can or should try to account for it because it would be too hard. Similar for the ringbuffers, although in that case it's much easier to predict how much data will come from cache vs from cassandra, but at the same time I think we want to discourage the static ringbuffers; so I think we can treat all "points that need to be loaded" the same.

Also, I suppose there's a big difference between requesting 10M points from C* and returning those 10M to the user vs requesting 10M points from C* but through sum() or runtime consolidation, only returning a tiny response back to the client.  Of course having a processing layer in between makes it hard to make predictions about response size, but I think this is worth pointing out: do we want to be able to prevent clients shooting themselves in the foot and erroring instead of trying to have them download a response that would take them many minutes?

Also the processing itself has a cost.  one will be able to load a relatively small data set but do a lot of processing on it, making the request expensive.  But I don't think we can or should account for this separately.

either way, I think the right answer will probably be set of different restrictions.

**Comment by woodsaj** [comment]
Changing the rollup used probably has a negligible impact on cassandra performance.
Lower resolution rollups could see an performance gain as fewer SSTABLES will need to be used.  But that is dependent on the size of the sstables used and the time span.
Generally i would expect the performance to the about the same as the same number of partitions will need to be read regardless of the rollup. The the volume of data will be different, but reading all of the data from a partition needed to serve a request should be performed with a single sequential read.

**Comment by Dieterbe** [comment]
> Lower resolution rollups could see an performance gain as fewer SSTABLES will need to be used. But that is dependent on the size of the sstables used and the time span.

but we tie number of sstables to TTL of the rollup. so the most accurate way to 
describe number of sstables used for a given series is "time range queried divided by TTL".
number of points is an appropriate approximation of this as long as proportion between TTL and interval is more or less constant across different rollup archives, this is often the case. but not always.for example in our ops HM,  `wp-usage` we retain 1h for 35d and 2h for 2y. setting a "time range queried divided by TTL restriction" such as 0.5 would mean "only query up to half of the sstables of the archive" (for either achive), but expressing this as a points limit such as 420 points for a series means only up to half of the sstables of the archive for the first archive (because 420 hourly points is half of the TTL), but for the second archive, 420 points, gets us 35 days which is only 0.048 of the timespan, e.g. of the sstables.
So number of points is sometimes not a good way to model, and hence to restrict, cassandra load.
As for number of partitions being queried, we currently don't have any provision to limit this. Due to our hash based partitioning though, the number of partitions used in a single query is arbitrary and not very related to amount of series being queried.

In other words, while number of points is a good way to model and restrict: 
A) tsz decoding workload for MT
B) decoding/processing workload for graphite-api

we currently do not have proper tunable resource limits for:
C) cassandra queries
D) responses returned back to the user. (consider that a typical graphite response contains for each point, something like `[8995.89, 1490823300], ` e.g. 23B per point. with the default `max-points-per-req-hard` of 20M introduced in #577 this means a response can be 460MB

and maybe this is OK, I think the main goal here was tackling B and not C/D (correct me if i'm wrong), C/D doesn't seem as urgent. Definitely D could be called pebkac. I just think it's worth being explicit about this.

**Comment by woodsaj** [comment]
yes, items A and B are the current bottle neck. Specifically B.

C can be constrained by adding a max targets limit.  But there is no urgency for that at this stage.
